### ISSUE TITLE
Research LayerNorm

### ISSUE THREAD

**Issue by CJcool06** [ChrisFuscoMasters/TransformerImpl#10]
**Aim**
Find out what layer-norm actually does (ie. benefits, limitations) and why/how it's applied to transformers.

**Plan**
- [ ] [Understanding the Difficulty of Training Transformers](https://arxiv.org/abs/2004.08249)
- [ ] [Understanding and Improving Layer Normalization](https://arxiv.org/abs/1911.07013)
- [ ] [On Layer Normalization in the Transformer Architecture](https://arxiv.org/abs/2002.04745)
- [ ] [PowerNorm: Rethinking Batch Normalization in Transformers](https://arxiv.org/abs/2003.07845)
- [ ] [ReZero is All You Need: Fast Convergence at Large Depth](https://arxiv.org/abs/2003.04887)
- [ ] [Characterizing signal propagation to close the performance gap in unnormalized ResNets](https://arxiv.org/abs/2101.08692)

**Comment by CJcool06** [comment]
**Tentative:**

[Understanding the Difficulty of Training Transformers](https://arxiv.org/abs/2004.08249) find that the mean and variance of layer normalisation (LN) are more important than the forward normalisation due to their re-centering and re-scaling of backward gradients. They also find that the parameters of LN increase the risk of overfitting and do not work in most cases. They propose Adaptive Normalisation (AdaNorm) that demonstrates better results than LN on 7/8 datasets.

[On Layer Normalization in the Transformer Architecture](https://arxiv.org/abs/2002.04745) show that the layer normalisation (LN) position in the vanilla Transformer causes the expected gradients of the parameters near the output layer to be large. Using a large learning rate on these gradients makes training unstable. They propose putting the LN inside of the residual blocks, before the multi-head attention and before the FFN. They verify that the training warm-up stage can be safely removed.
